Support receiving to multi-hop blinded paths

valentinewallace opened the pull request:
Title. Currently, we only support receiving to 1-hop blinded paths. Diff is largely error handling. 

- [x] Add test coverage for malformed HTLC ser

~Based on #2540.~

valentinewallace commented on lightning/src/ln/channel.rs:
```diff
@@ -259,6 +259,9 @@ enum HTLCUpdateAwaitingACK {
 		htlc_id: u64,
 		err_packet: msgs::OnionErrorPacket,
 	},
+	FailBlindedHTLC {
```
Yeah, good point. If that's the ultimate goal though, I think we should write these failed HTLCs as a new TLV in `Channel` rather than always failing with an empty error packet per your other ser comment.

valentinewallace replied:
Ah yes, we should also write the empty error packet 👍 

Been working on this and adding the TLV on top adds a bit of complexity, unfortunately :( It feels like the right thing to do though.

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -9569,6 +9631,9 @@ impl_writeable_tlv_based_enum!(HTLCForwardInfo,
 		(0, htlc_id, required),
 		(2, err_packet, required),
+	},
+	(2, FailBlindedHTLC) => {
```
I think this breaks compat too, unfortunately :( Will update.

valentinewallace commented:
Addressed feedback. Planning to add some test coverage for the new malformed HTLC ser changes, added a TODO.

valentinewallace commented:
Rebased.

valentinewallace commented on lightning/src/ln/channel.rs:
```diff
@@ -7470,9 +7474,17 @@ impl<SP: Deref> Writeable for Channel<SP> where SP::Target: SignerProvider {
+					2u8.write(writer)?;
 					htlc_id.write(writer)?;
+					dummy_err_packet.write(writer)?;
 				}
```
No, downgraded nodes will use this dummy htlc data to fail this htlc backwards. This prevents them from dropping the malformed HTLCs (which are written as optional TLVs) on `read` post-downgrade, which would cause them to not fail backwards and eventually be FC'd on.

valentinewallace commented on lightning/src/ln/channel.rs:
```diff
@@ -7766,13 +7779,16 @@ impl<'a, 'b, 'c, ES: Deref, SP: Deref> ReadableArgs<(&'a ES, &'b SP, u32, &'c Ch
+
+					// An empty error packet indicates we wrote this `::FailHTLC` for compat, and the real
+					// malformed error will be retrieved from the TLVs.
+					if err_packet.data.is_empty() { continue }
```
In the future when we move to deserializing the onion in `process_pending_htlc_forwards`, we'll write an empty `OnionErrorPacket` for all malformed HTLCs, not just blinded ones. 

In general, there shouldn't be any spec-valid reason to fail backwards with an empty error packet, so unless a new htlc failure message is added in the spec that we want to support downgrades for I don't see why we would ever create an empty one. @TheBlueMatt may want to double check.

valentinewallace commented on lightning/src/ln/channel.rs:
```diff
@@ -2863,22 +2863,11 @@ impl<SP: Deref> Channel<SP> where
-	/// want to fail blinded HTLCs where we are not the intro node.
-	///
-	/// See [`Self::queue_fail_htlc`] for more info.
-	pub fn queue_fail_malformed_htlc<L: Deref>(
```
To get rid of this method, I had to increase the visibility of some internal `Channel` structs. Might be better to just keep a separate method for failing malformed HTLCs, let me know if there are thoughts.

valentinewallace commented on lightning/src/ln/channel.rs:
```diff
@@ -2871,6 +2871,18 @@ impl<SP: Deref> Channel<SP> where
+			htlc_id_arg, (failure_code, sha256_of_onion), true, logger
+		).map(|msg_opt| assert!(msg_opt.is_none(), "We forced holding cell?"))
+	}
+
```
Yeah, although see https://github.com/lightningdevkit/rust-lightning/pull/2688#discussion_r1423054495.

valentinewallace replied:
Re-duplicated after offline discussion since it's more readable in `ChannelManager` this way.

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -4387,8 +4387,19 @@ where
+										// pending, and if not that's OK, if not, the channel is on
+										// the chain and sending the HTLC-Timeout is their problem.
+										continue;
+									}
```
I don't think so, because `queue_fail_htlc` statically chooses the concrete underlying type of `FailHTLCContents` at compile time, IIUC.

jkczyz commented on lightning/src/ln/channel.rs:
```diff
@@ -2518,6 +2523,52 @@ struct CommitmentTxInfoCached {
+/// Contents of a wire message that fails an HTLC backwards. Useful for [`Channel::fail_htlc`] to
+/// fail with either [`msgs::UpdateFailMalformedHTLC`] or [`msgs::UpdateFailHTLC`] as needed.
+pub(super) trait FailHTLCContents {
+	type Message;
+	fn message(self, htlc_id: u64, channel_id: ChannelId) -> Self::Message;
+	fn message_name() -> &'static str;
+	fn inbound_htlc_state(self) -> InboundHTLCState;
+	fn htlc_update_awaiting_ack(self, htlc_id: u64) -> HTLCUpdateAwaitingACK;
```
Somewhat pedantic, but:
- naming could use `to_` prefix for methods taking `self`
- `message_name` arguably belongs on it's own trait (`MessageName`?) where the `Message` associated type has a trait bound on it.

jkczyz commented on lightning/src/ln/channel.rs:
```diff
@@ -2518,6 +2523,52 @@ struct CommitmentTxInfoCached {
+	}
+	fn message_name() -> &'static str { "update_fail_malformed_htlc" }
+	fn inbound_htlc_state(self) -> InboundHTLCState {
+		let (failure_code, sha256_of_onion) = self;
```
Should we normalize the order so this doesn't need to be destructured?

valentinewallace replied:
Got rid of the destructuring. Not sure I got what you meant by "normalize."

jkczyz replied:
By that I meant to keep the order of `(failure_code, sha256_of_onion)` the same between here and in `InboundHTLCRemovalReason::FailMalformed`. That way you just need to wrap `self` here rather than accessing each item in the tuple.

valentinewallace replied:
Oh, okay. I'll pass on this for now but I'll do it if there ends up being a follow-up to this PR.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -6558,14 +6618,24 @@ where
 								ref incoming_shared_secret, ref routing, ..
 							}) => {
-								let reason = if routing.blinded_failure().is_some() {
+								let reason =  if routing.blinded_failure().is_some() {
```
nit: extra space

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -6558,14 +6618,24 @@ where
 						Err(e) => PendingHTLCStatus::Fail(e)
 					};
 					let create_pending_htlc_status = |chan: &Channel<SP>, pending_forward_info: PendingHTLCStatus, error_code: u16| {
+						if msg.blinding_point.is_some() {
+							return PendingHTLCStatus::Fail(HTLCFailureMsg::Malformed(
+									msgs::UpdateFailMalformedHTLC {
+										channel_id: msg.channel_id,
+										htlc_id: msg.htlc_id,
+										sha256_of_onion: [0; 32],
+										failure_code: INVALID_ONION_BLINDING,
+									}
+							))
+						}
```
Just want to check that the update here isn't specific to the `PendingHTLCStatus::Forward` case in the below `match` as was done in the previous version.

jkczyz commented on lightning/src/ln/channel.rs:
```diff
@@ -8032,6 +8138,22 @@ impl<'a, 'b, 'c, ES: Deref, SP: Deref> ReadableArgs<(&'a ES, &'b SP, u32, &'c Ch
 		}
 
+		if let Some(malformed_htlcs) = malformed_htlcs {
+			for (htlc_id, failure_code, sha256_of_onion) in malformed_htlcs {
+				let dummy_htlc_idx = holding_cell_htlc_updates.iter().position(|htlc| {
+					if let HTLCUpdateAwaitingACK::FailHTLC { htlc_id: dummy_htlc_id, err_packet } = htlc {
+						let matches = htlc_id == *dummy_htlc_id;
+						if matches { debug_assert!(err_packet.data.is_empty()) }
+						matches
+					} else { false }
+				}).ok_or(DecodeError::InvalidValue)?;
+				let malformed_htlc = HTLCUpdateAwaitingACK::FailMalformedHTLC {
+					htlc_id, failure_code, sha256_of_onion
+				};
+				let _ = core::mem::replace(&mut holding_cell_htlc_updates[dummy_htlc_idx], malformed_htlc);
```
I'd drop the `dummy_` prefix throughout and simply rename `htlc_id` to `malformed_htlc_id`. It's only a dummy htlc for ones that match.

jkczyz reviewed (commented):
I'd be fine with a squash now. Just need to take a closer look at the tests.

valentinewallace commented:
Squashed and addressed feedback with 1 new fixup.

jkczyz reviewed (commented):
LGTM

valentinewallace commented on lightning/src/ln/channel.rs:
```diff
@@ -3563,6 +3637,20 @@ impl<SP: Deref> Channel<SP> where
 					},
+					&HTLCUpdateAwaitingACK::FailMalformedHTLC { htlc_id, failure_code, sha256_of_onion } => {
+						match self.fail_htlc(htlc_id, (failure_code, sha256_of_onion), false, logger) {
+							Ok(update_fail_malformed_opt) => {
```
To address this and the below, I believe I'll need to introduce a macro or helper function, as I'm getting:

```rust
error[E0408]: variable `err_packet` is not bound in all patterns
    --> lightning/src/ln/channel.rs:3622:6
     |
3621 |                     &HTLCUpdateAwaitingACK::FailHTLC { htlc_id, ref err_packet } |
     |                                                                     ---------- variable not in all patterns
3622 |                     &HTLCUpdateAwaitingACK::FailMalformedHTLC { htlc_id, failure_code, sha256_of_onion } => {
     |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ pattern doesn't bind `err_packet`
```

Unless there's a way to make a tuple out of the two malformed variant args in the match arm itself?

valentinewallace commented:
Squashed the fixup and added another commit addressing https://github.com/lightningdevkit/rust-lightning/pull/2688#discussion_r1424608874. Sorry if I'm being dense on DRYing the match arms, but happy to address that and https://github.com/lightningdevkit/rust-lightning/pull/2688#discussion_r1423190055 in a follow-up.
